src/digest/sha2.rs

```diff
@@ -0,0 +1,248 @@
+// Copyright 2015-2016 Brian Smith.
+// Copyright 2016 Simon Sapin.
+// Copyright 2016 Sam Scott.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use c;
+use core;
+use core::num::Wrapping;
+use polyfill;
+use std::ops::{Add,BitAnd,BitXor,Shr,Not};
+
+use super::MAX_CHAINING_LEN;
+
+// SHA 224/256: 512-bit blocks
+const BLOCK_LEN_256: usize = 512/8;
+
+// SHA 384/512: 1024-bit blocks
+const BLOCK_LEN_512: usize = 1024/8;
+
+// SHA 224/256: State is 256 bits
+pub const CHAINING_LEN_256: usize = 256 / 8;
+// SHA 384/512: State is 512 bits
+pub const CHAINING_LEN_512: usize = 512 / 8;
+
+// State as number of words
+const CHAINING_WORDS_256: usize = CHAINING_LEN_256 / 4;
+const CHAINING_WORDS_512: usize = CHAINING_LEN_512 / 8;
+
+
+type W32 = Wrapping<u32>;
+type W64 = Wrapping<u64>;
+
+trait Word: Shr<usize,Output=Self> + Add<Self,Output=Self> + BitAnd<Self,Output=Self> +
+            BitXor<Self,Output=Self> + Not<Output=Self> + Copy + Sized {
+    #[inline]
+    fn ch(x: Self, y: Self, z: Self) -> Self {
+        (x & y) ^ (!x & z)
+    }
+    #[inline]
+    fn parity(x: Self, y: Self, z: Self) -> Self {
+        x ^ y ^ z
+    }
+    #[inline]
+    fn maj(x: Self, y: Self, z: Self) -> Self {
+        (x & y) ^ (x & z) ^ (y & z)
+    }
+
+    // Performs rotate right
+    #[inline]
+    fn rotr(self, n: u32) -> Self;
+}
+
+impl Word for W32 {
+    #[inline]
+    fn rotr(self, n: u32) -> W32 {
+        Wrapping(self.0.rotate_right(n))
+    }
+}
```

Comment:
I guess if we do it this way, we should update the SHA-1 code to follow the same pattern, right?